README.md
Scallion
========
Scallion lets you make create .onion addresses for [Tor's](https://www.torproject.org/) [hidden services](https://www.torproject.org/docs/hidden-services) using OpenCL. It is based in part on [shallot](https://github.com/katmagic/Shallot) and inspired by [vanitygen](https://github.com/samr7/vanitygen). The OpenCL SHA-1 implementation was adapted from [NearSHA](http://cr.yp.to/nearsha.html).

Scallion runs on Mono (tested in Arch Linux) and .NET 3.5 (tested on Windows 7 and Server 2008). It includes components from [OpenTK](http://www.opentk.com/) and [OpenSSL-net](http://openssl-net.sourceforge.net/).

Scallion is currently in alpha stage and under active development. Nevertheless, we feel that it is ready for use in generating tor hidden service private. Improvements are expected primarily in performance, user interface, and ease of installation, not in the overall algorithm used to generate keys.

Scallion is available under the MIT licence. If you use code from this project in your own projects, we'd appreciate hearing about it at scallion@aftbit.com.

Dependancies
------------
- OpenCL and relevant drivers installed and configured. Refer to your distribution's documentation.
- OpenSSL. For Windows, the prebuilt x86 DLLs are included

Build Linux
-----------
    xbuild scallion.sln

Build Windows
-------------
1. Open 'scallion.sln' in Visual Studio
2. Build the solution

Usage
-----
List devices

    $ mono scallion/bin/Debug/scallion.exe -l

Generate a hash

    $ mono scallion/bin/Debug/scallion.exe -d 0 prefix
    Cooking up some delicions scallions...
    LoopIteration:15  HashCount:251.66MH  Speed:89.2MH/s  Runtime:00:00:02  Predicted:00:00:12
    Ding!! Delicions scallions for you!!

    Exponent: 37074435
    Address/Hash: prefix2bp7lfuuvp

    -----BEGIN RSA PRIVATE KEY-----
    MIICXQIBAAKBgQDVNxlMDVXQ6EjRLubgMUkhVeVYigEPZ4BLUhzNRp4MEMgVQHLP
    GRlMc2yK29Q8fuvC1o2zJS8IF6RbXyB9Sdyuzh43st2CZeTMEWbkz6NNAJz+8UHh
    1I35CWx5p4wlsw2eZx+wM7s6Ll4762pV21qolxqHoIefOsIso0AHixYPrQIEAjW2
    AwKBgAEqghqEMZ2cedXc+AIKmZebbzJyWvfp9W9HRHXn6c7U0mYFNHXnAjR8KR6r
    2w2IGS4LxKi360XRr70gIUw9mr9tiOlMjppkqwu7HSB0ldzCitNkVLRiV+TZQFXc
    g6xxZZxX2giZoBThCq8g85/V+AyLRZK9ZdC+GkBP0YPgy8/rAkEA/j8vGpo/OkZq
    3ucytAZYb+1HomUiSbh+oxMwxT84xQeSqIg/BsvWixPhuOY+7HNzdZVN458H8UXC
    zSWNmmmnYQJBANave5NDwnw5w/dARWMJsAYRc7GA/wx86o/+qmw/8Q6GBkFdUqcM
    2Vw0HzDIq7q1UwYBznlRCI1Wgyd5+OwDZ80CQQDaewQ811o3/8StlKLvpify+fkQ
    81j0GdoUJgYCz3nDEp6sCPvg3aSI7b195odY4L3d0pQ4SnPj0zGJMFdqcwFLAkBj
    A/NC23ZfVx8u1JzRojkwuE4ZAWmALIubP+iwS6I5Yj4/wz2R1veYBQX1TKNT8sOy
    XA8cQ53ybVp+z39eWCPHAkAz1Sv1IOuGWS/KrC0BUQ7vtwf3z5yMUX3kMS1JF9Bf
    TbEyV92GNkwxcVzMRCSeaqZXQVUIaWdbPnjYq4T6Heal
    -----END RSA PRIVATE KEY-----


    init: 423ms / 1 (423ms, 2.36/s)
    generate key: 840ms / 45 (18.67ms, 53.57/s)
    cpu precompute: 9ms / 45 (0.2ms, 5000/s)
    total without init: 2828ms / 1 (2828ms, 0.35/s)
    set buffers: 0ms / 15 (0ms, 0/s)
    write buffers: 2ms / 15 (0.13ms, 7500/s)
    run kernel: 0ms / 15 (0ms, 0/s)
    read results: 2812ms / 15 (187.47ms, 5.33/s)
    check results: 5ms / 15 (0.33ms, 3000/s)

    88.99 million hashes per second

Speed
-----
On my nVidia Quadro K2000M, I see around 90MH/s. With those speeds, I can generate a six character prefix in about six seconds on average. An eight character prefix would take about 1h 45m. To calculate the number of seconds required for a given prefix (on average), use this formula:

    seconds = 2^(5*length-1) / hashspeed 

Security
--------
The keys generated by Scallion are quite similar to those generated by shallot. They have unusually large public exponents, but they are put through the full set of sanity checks recommended by PKCS #1 v2.1 via openssl's RSA_check_key function. In general, however, tor's (and Scallion's) default 1024-bit key is starting to [look a little small](https://lists.torproject.org/pipermail/tor-dev/2011-November/003033.html). Scallion supports several RSA key sizes, with optimized kernels for 1024b, 2048b, and 4096b. Other key sizes may work, but have not been tested.



openssl-net-0.5/CHANGES
OpenSSL.NET - CHANGES

0.5 [Mar 29, 2012]

* Breaking change: SSL API has been moved to its own namespace.
* Bug fixes reported on sourceforge.
* Adding Elliptic Curve support.
* Conversion of tests to NUnit.

0.5-rc2 [Feb 26, 2012]

* Updating linux and osx native libraries to version 1.0.0g
* Working around a few 64-bit issues found on Linux under Mono.

0.5-rc1 [Apr 14, 2011]

* Updating OpenSSL native library to version 1.0.0d
* Using Visual Studio 2010 project files 
  (I don't have the old versions of VS installed anymore)
* Removing EVP_md2(), the native library no longer includes this by default

0.4.4 [Jan 18, 2010]

* Added support for i2d_X509_bio() which allows for writing X509
  certificates in DER format.

0.4.3 [Oct 23, 2009]

* Added support for PKCS12_create().

0.4.2 [Oct 23, 2009]

* Replacing the prebuilt binaries for libeay32.dll and ssleay32.dll with
  ones that come directly from the openssl source. This should fix
  problems people are finding in the wild, but it will also require
  them to install the MSVC 2008 CRT first.

0.4.1 [Sep 9, 2009]

* Including prebuilt binaries to libeay32.dll and ssleay32.dll that are
  statically linked to the CRT instead of depending on the DLL CRT. 
  This should resolve the DllNotFoundException at startup.

0.4 [Aug 26, 2009]

* Thanks to Ben Henderson for contributing the wrapper for the SSL API!
* Lots of bug fixes mostly concerning memory management.
* Refactored the Base class to deal with memory mangement better.
* Default project build now produces assemblies specifically targetted for
  the x86 architecture, which works around issues with using the native
  DLL on 64-bit machines.
* Lots more XML comments/docs (no more compiler warnings!)
* Changes to namespace on most classes so it's easier to navigate them.
* Added unit tests for X509Certificate and CryptoKey (many more to go)
* Support for CE has been dropped due to lack of testing. Will add again
  once I have the ability to test it again.
* Added HMAC wrapper.

0.3.2 [Jun 01, 2008]

* Added wrapper and test for EVP_BytesToKey().

0.3.1 [May 11, 2008]

* Thanks to Christoph Scrauth for helping me find and fix a few bugs.

0.3 [Jan 02, 2008]

* Added RAND wrapper and tests.
* Added RSA wrapper and tests.
* Added a couple more commands to openssl+.exe (genrsa, rsa)
* Added comments to all publicly facing interfaces.

0.2 [Dec 20, 2007]

* Fixed callbacks to use __cdecl instead of the default of __stdcall.
* Added SHA tests.
* Added DH tests.
* Upgraded the DSA wrapper.
* Upgraded the DH wrapper.

0.1 [Apr 21, 2007]

* Initial public release

scallion/CLRuntime.cs
using System;
using System.Linq;
using System.Collections.Generic;
using OpenSSL.Core;
using OpenSSL.Crypto;
using System.Threading;

namespace scallion
{
	public enum KernelType
	{
		Normal,
		Optimized4_9,
		Optimized4_11
	}

	public class CLRuntime
	{
		public static List<CLDeviceInfo> GetDevices()
		{
			return CLDeviceInfo.GetDeviceIds()
				.Select(i => new CLDeviceInfo(i))
				.Where(i => i.CompilerAvailable)
				.ToList();
		}

		private KernelType kernel_type;
		private int keySize;

		static private int get_der_len(ulong val)
		{
			if(val == 0) return 1;
			ulong tmp = val;
			int len = 0;

			// Find the length of the value
			while(tmp != 0) {
				tmp >>= 8;
				len++;
			}

			// if the top bit of the number is set, we need to prepend 0x00
			if(((val >> 8*(len-1)) & 0x80) == 0x80)
				len++;

			return len;
		}
		
		public class KernelInput
		{
			public KernelInput(KernelInput input, uint baseExp)
			{
				Rsa = input.Rsa;
				LastWs = input.LastWs;
				Midstates = input.Midstates;
				ExpIndexes = input.ExpIndexes;
				Results = input.Results;
				BaseExp = baseExp;
			}
			public KernelInput(int num_exps)
			{
				Rsa = new RSAWrapper();
				LastWs = new uint[num_exps * 16];
				Midstates = new uint[num_exps * 5];
				ExpIndexes = new int[num_exps];
				Results = new uint[128];
				BaseExp = EXP_MIN;
			}
			public readonly uint[] LastWs;
			public readonly uint[] Midstates;
			public readonly int[] ExpIndexes;
			public readonly RSAWrapper Rsa;
			public readonly uint[] Results;
			public readonly uint BaseExp;
		}
		const uint EXP_MIN = 0x01010001;
		const uint EXP_MAX = 0x7FFFFFFF;
		public bool Abort = false;
		private Queue<KernelInput> _kernelInput = new Queue<KernelInput>();
		private void CreateInput()
		{
			while (true)
			{
				bool inputQueueIsLow = false;
				lock (_kernelInput)	{ inputQueueIsLow = _kernelInput.Count < 300; }
				if (inputQueueIsLow)
				{
					int num_exps = (get_der_len(EXP_MAX) - get_der_len(EXP_MIN) + 1);
					KernelInput input = new KernelInput(num_exps);

					profiler.StartRegion("generate key");
					input.Rsa.GenerateKey(keySize); // Generate a key
					profiler.EndRegion("generate key");

					// Build DERs and calculate midstates for exponents of representitive lengths
					profiler.StartRegion("cpu precompute");
					int cur_exp_num = 0;
					BigNumber[] Exps = new BigNumber[num_exps];
					bool skip_flag = false;
					for (int i = get_der_len(EXP_MIN); i <= get_der_len(EXP_MAX); i++)
					{
						ulong exp = (ulong)0x01 << (int)((i - 1) * 8);

						// Set the exponent in the RSA key
						// NO SANITY CHECK - just for building a DER
						input.Rsa.Rsa.PublicExponent = (BigNumber)exp;
						Exps[cur_exp_num] = (BigNumber)exp;

						// Get the DER
						byte[] der = input.Rsa.DER;
						int exp_index = der.Length % 64 - i;
						if (kernel_type == KernelType.Optimized4_9) {
							if(exp_index != 9) { // exponent index assumed to be 9 in the kernel
								Console.WriteLine("Exponent index doesn't match - skipping key");
								skip_flag = true;
								break;
							}
							if(i != 4) { // exponent length assumed to be 4 in the kernel
								Console.WriteLine("Exponent length doesn't match - skipping key");
								skip_flag = true;
								break;
							}
						}
						else if (kernel_type == KernelType.Optimized4_11) {
							if(exp_index != 11) { // exponent index assumed to be 9 in the kernel
								Console.WriteLine("Exponent index doesn't match - skipping key");
								skip_flag = true;
								break;
							}
							if(i != 4) { // exponent length assumed to be 4 in the kernel
								Console.WriteLine("Exponent length doesn't match - skipping key");
								skip_flag = true;
								break;
							}
						}

						// Put the DER into Ws
						SHA1 Sha1 = new SHA1();
						List<uint[]> Ws = Sha1.DataToPaddedBlocks(der);

						// Put all but the last block through the hash
						Ws.Take(Ws.Count - 1).Select((t) =>
						{
							Sha1.SHA1_Block(t);
							return t;
						}).ToArray();

						// Put the midstate, the last W block, and the byte index of the exponent into the CL buffers
						Sha1.H.CopyTo(input.Midstates, 5 * cur_exp_num);
						Ws.Last().Take(16).ToArray().CopyTo(input.LastWs, 16 * cur_exp_num);
						input.ExpIndexes[cur_exp_num] = exp_index;

						// Increment the current exponent size
						cur_exp_num++;
						break;
					}
					profiler.EndRegion("cpu precompute");

					if(skip_flag) continue; // we got a bad key - don't enqueue it

					List<KernelInput> inputs = new List<KernelInput>();
					inputs.Add(input);
					for (uint i = 1; i < (EXP_MAX - EXP_MIN) / 4 / workSize - 1; i++)
					{
						profiler.StartRegion("generate key");
						inputs.Add(new KernelInput(input, EXP_MIN + workSize * i));
						profiler.EndRegion("generate key");
					}
					lock (_kernelInput)//put input on queue
					{
						foreach (KernelInput i in inputs)
						{
							_kernelInput.Enqueue(i);
						}
					}
					continue;//skip the sleep cause we might be really low
				}
				Thread.Sleep(50);
			}
		}

		private TimeSpan PredictedRuntime(string prefix, string suffix, long speed)
		{
			int len = prefix.Length + suffix.Length;
			long runtime_sec = (long)Math.Pow(2,5*len-1) / speed;
			int days=(int)(runtime_sec/86400), hrs=(int)((runtime_sec%86400)/3600), min=(int)(runtime_sec%3600)/60, sec=(int)(runtime_sec%60);
			TimeSpan ts = new TimeSpan(days,hrs,min,sec);
			return ts;
		}

		private Profiler profiler = null;
		private uint workSize; 

		public void Run(int deviceId, int workGroupSize, int workSize, KernelType kernelt, int keysize, string prefix, string suffix)
		{
			Console.WriteLine("Cooking up some delicions scallions...");
			this.workSize = (uint)workSize;
			profiler = new Profiler();
			#region init
			profiler.StartRegion("init");

			// Set the key size
			keySize = keysize;

			// Find kernel name and check key size
			kernel_type = kernelt;
			string kernelFileName=null, kernelName=null;
			switch (kernel_type) {
				case KernelType.Normal:
					kernelFileName = "kernel.cl";
					kernelName = "normal";
					break;
				case KernelType.Optimized4_9:
					if(keySize != 1024) throw new ArgumentException("Kernel {0} only works with keysize 1024.");	
					kernelFileName = "kernel.cl";
					kernelName = "optimized4_9";
					break;
				case KernelType.Optimized4_11:
					if(keySize != 2048 && keySize != 4096) throw new ArgumentException("Kernel {0} only works with keysize 2048 or 4096.");	
					kernelFileName = "kernel.cl";
					kernelName = "optimized4_11";
					break;
				default:
					throw new ArgumentException("Pick a supported kernel.");
			}

			Console.WriteLine("Using kernel {0} from file {1} ({2})",kernelName,kernelFileName,kernel_type);

			//create device context and kernel
			CLDeviceInfo device = GetDevices()[deviceId];
			if ((uint)workGroupSize > device.MaxWorkGroupSize)
			{
				workGroupSize = (int)device.MaxWorkGroupSize;
			}
			CLContext context = new CLContext(device.DeviceId);
			IntPtr program = context.CreateAndCompileProgram(
				System.IO.File.ReadAllText(
					System.IO.Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location) + System.IO.Path.DirectorySeparatorChar + kernelFileName
				)
			);

			CLKernel kernel = context.CreateKernel(program, kernelName);
			//Create buffers
			CLBuffer<uint> bufLastWs;
			CLBuffer<uint> bufMidstates;
			CLBuffer<int> bufExpIndexes;
			CLBuffer<uint> bufResults;
			{
				int num_exps = (get_der_len(EXP_MAX) - get_der_len(EXP_MIN) + 1);
				uint[] LastWs = new uint[num_exps * 16];
				uint[] Midstates = new uint[num_exps * 5];
				int[] ExpIndexes = new int[num_exps];
				uint[] Results = new uint[128];

				bufLastWs = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadOnly | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, LastWs);
				bufMidstates = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadOnly | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, Midstates);
				bufExpIndexes = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadOnly | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, ExpIndexes);
				bufResults = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadWrite | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, Results);
			}
			//Create pattern buffers
			CLBuffer<uint> bufPattern;
			CLBuffer<uint> bufBitmask;
			{
				string patternStr = prefix + "".PadLeft(16 - prefix.Length - suffix.Length, 'a') + suffix;
				uint[] Pattern = TorBase32.ToUIntArray(TorBase32.FromBase32Str(patternStr));
				string bitmaskStr = "".PadLeft(prefix.Length, 'x') + "".PadLeft(16 - prefix.Length - suffix.Length, '_') + "".PadLeft(suffix.Length, 'x');
				uint[] Bitmask = TorBase32.ToUIntArray(TorBase32.CreateBase32Mask(bitmaskStr));
				bufPattern = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadOnly | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, Pattern);
				bufBitmask = context.CreateBuffer(OpenTK.Compute.CL10.MemFlags.MemReadOnly | OpenTK.Compute.CL10.MemFlags.MemCopyHostPtr, Bitmask);
			}
			//Set kernel arguments
			lock (new object()) { } // Empty lock, resolves (or maybe hides) a race condition in SetKernelArg
			kernel.SetKernelArg(0, bufLastWs);
			kernel.SetKernelArg(1, bufMidstates);
			kernel.SetKernelArg(2, bufExpIndexes);
			kernel.SetKernelArg(3, bufResults);
			kernel.SetKernelArg(4, (uint)EXP_MIN);
			kernel.SetKernelArg(5, (byte)get_der_len(EXP_MIN));
			kernel.SetKernelArg(6, bufPattern);
			kernel.SetKernelArg(7, bufBitmask);
			profiler.EndRegion("init");

			//start the thread to generate input data
			Thread inputThread = new Thread(CreateInput);
			inputThread.Start();
			Thread.Sleep(3000);//wait just a bit so some work is available
			#endregion

			int loop = 0;

			var gpu_runtime_sw = System.Diagnostics.Stopwatch.StartNew();

			profiler.StartRegion("total without init");
			bool success = false;
			while (!success)
			{
				lock (this) { if (this.Abort) break; } //abort flag was set.... bail
				KernelInput input = null;
				lock (_kernelInput)
				{
					if (_kernelInput.Count > 0) input = _kernelInput.Dequeue();
				}
				if (input == null) //If we have run out of work sleep for a bit
				{
					Console.WriteLine("Lack of work for the GPU!! Taking a nap!!");
					Thread.Sleep(250);
					continue;
				}

				profiler.StartRegion("set buffers");
				bufLastWs.Data = input.LastWs;
				bufMidstates.Data = input.Midstates;
				bufExpIndexes.Data = input.ExpIndexes;
				bufResults.Data = input.Results;
				kernel.SetKernelArg(4, input.BaseExp);
				profiler.EndRegion("set buffers");

				profiler.StartRegion("write buffers");
				bufLastWs.EnqueueWrite();
				bufMidstates.EnqueueWrite();
				bufExpIndexes.EnqueueWrite();
				bufResults.EnqueueWrite();
				profiler.EndRegion("write buffers");

				profiler.StartRegion("run kernel");
				//kernel.EnqueueNDRangeKernel(1024*1024*16,128);
				kernel.EnqueueNDRangeKernel(workSize, workGroupSize);
				profiler.EndRegion("run kernel");

				profiler.StartRegion("read results");
				bufResults.EnqueueRead();
				profiler.EndRegion("read results");

				loop++;
				Console.Write("\r");
				long hashes = (long)workSize * (long)loop;
				Console.Write("LoopIteration:{0}  HashCount:{1:0.00}MH  Speed:{2:0.0}MH/s  Runtime:{3}  Predicted:{4}", 
				              loop, hashes / 1000000.0d, hashes/gpu_runtime_sw.ElapsedMilliseconds/1000.0d, 
				              gpu_runtime_sw.Elapsed.ToString().Split('.')[0], 
				              PredictedRuntime(prefix,suffix,hashes*1000/gpu_runtime_sw.ElapsedMilliseconds));

				profiler.StartRegion("check results");
				foreach (var result in input.Results)
				{
					if (result != 0)
					{
						try
						{
							Console.WriteLine();
							Console.WriteLine("Ding!! Delicions scallions for you!!");
							Console.WriteLine();
							Console.WriteLine("Exponent: {0}", result);
							input.Rsa.ChangePublicExponent((BigNumber)result);
							Console.WriteLine("Address/Hash: " + input.Rsa.OnionHash);
							Console.WriteLine();
							Console.WriteLine(input.Rsa.Rsa.PrivateKeyAsPEM);
							Console.WriteLine();
							success = true;
						}
						catch (OpenSslException /*ex*/) { }
					}
				}
				profiler.EndRegion("check results");
			}

			inputThread.Abort();//stop makin work
			profiler.EndRegion("total without init");
			Console.WriteLine(profiler.GetSummaryString());
			Console.WriteLine("{0:0.00} million hashes per second", ((long)loop * (long)workSize * (long)1000) / (double)profiler.GetTotalMS("total without init") / (double)1000000);
		}
	}
}

